### PR TITLE
Manual ACK on QoS 1 and 2

### PR DESCRIPTION
**Pull request by dometec**
2nd PR

Added AutoAck option (default true)

## old description

The problem I found in VertX MQTT Client implementation is that acknowledgment of the message is done by the library before passing the message to the application logic. So the application does not have the ability to mark the message as elaborated. In this way, MQTT messages at QoS1 or QoS2 can be lost before they can be elaborated, for example for a crash of the client, since the broker has received the acks for those messages, and for it, are successfully delivered.

I extended the message handler of the VertX client and I give the opportunity to application logic to ack the message.

A more detailed example is there: https://github.com/dometec/vertx-mqtt-flowcontrol-example

I'm sorry to not provide a unit test for it, but I think that it's not so easy...

### CODE REVIEW

File: src/main/java/io/vertx/mqtt/MqttClient.java [hunk 1]
```diff
@@ -180,13 +180,13 @@ static MqttClient create(Vertx vertx) {
...
    * Sets handler which will be called each time server publish something to client
-   *
+   * 
    * @param publishHandler handler to call
    * @return current MQTT client instance
    */
   @Fluent
   MqttClient publishHandler(Handler<MqttPublishMessage> publishHandler);
-
+
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttClient.java:189]: I think nothing is changed in this file, can you avoid this?

File: src/main/java/io/vertx/mqtt/MqttClientOptions.java [hunk 2]
```diff
@@ -371,6 +374,14 @@ public MqttClientOptions setAutoGeneratedClientId(boolean isAutoGeneratedClientI
     this.isAutoGeneratedClientId = isAutoGeneratedClientId;
     return this;
   }
+ 
+  public boolean isAutoAck() {
+	return autoAck;
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttClientOptions.java:379]: indentation

File: src/main/java/io/vertx/mqtt/MqttClientOptions.java [hunk 3]
```diff
@@ -371,6 +374,14 @@ public MqttClientOptions setAutoGeneratedClientId(boolean isAutoGeneratedClientI
...
+  }
+
+  public void setAutoAck(boolean autoAck) {
+	this.autoAck = autoAck;
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttClientOptions.java:383]: indentation

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 4]
```diff
@@ -507,7 +508,7 @@ public MqttClient publishHandler(Handler<MqttPublishMessage> publishHandler) {
   private synchronized Handler<MqttPublishMessage> publishHandler() {
     return this.publishHandler;
   }
-
+
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java:511]: avoid removing and adding line

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 5]
```diff
@@ -1173,30 +1174,35 @@ private void handleSuback(MqttSubAckMessage msg) {
...
+      	if (handler != null) {
+	        handler.handle(msg);
+      	}
         break;
 
+      case AT_LEAST_ONCE:  
+		  if (options.isAutoAck()) {			  
+			  this.publishAcknowledge(msg.messageId());
+		  } else {
+              ((MqttPublishMessageImpl) msg).setAckCallback(() -> this.publishAcknowledge(msg.messageId()));
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java:1192]: indentation

File: src/main/java/io/vertx/mqtt/messages/MqttPubAckCallback.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package io.vertx.mqtt.messages;
+
+public interface MqttPubAckCallback {
+
+	void ack();
+	
+}
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/messages/MqttPubAckCallback.java:7]: missing Javadoc on both interface declaration and method

**Comment by dometec** [comment]
I moved the MqttPubAckCallback interface into MqttPublishMessageImpl, because, even if it's an interface, it's something inside the library. Let me know what you think.

File: src/main/java/io/vertx/mqtt/messages/MqttPublishMessage.java [hunk 7]
```diff
@@ -94,6 +95,11 @@ static MqttPublishMessage create(int messageId, MqttQoS qosLevel, boolean isDup,
    */
   @CacheReturn
   Buffer payload();
+  
+  /**
+   * Send the PUBACK/PUBCOMP to the broker. Use this method only if autoAck option is set to false.
+   */
+  void ack();
```
Comment (vietj) [inline src/main/java/io/vertx/mqtt/messages/MqttPublishMessage.java:102]: this should throw IllegalStateException if ack was already performed (either once or automatically)

Comment (dometec) [reply src/main/java/io/vertx/mqtt/messages/MqttPublishMessage.java:104]: good point, I also remove the log and throw an IllegalStateException if you ack a message when Auto Ack is true

File: src/main/java/io/vertx/mqtt/MqttClientOptions.java [hunk 8]
```diff
@@ -371,6 +374,14 @@ public MqttClientOptions setAutoGeneratedClientId(boolean isAutoGeneratedClientI
...
     return this;
   }
+ 
+  public boolean isAutoAck() {
+     return autoAck;
+  }
+
+  public void setAutoAck(boolean autoAck) {
+    this.autoAck = autoAck;
+  }
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/mqtt/MqttClientOptions.java:393]: ping @dometec

File: src/main/java/io/vertx/mqtt/messages/impl/MqttPublishMessageImpl.java [hunk 9]
```diff
@@ -20,20 +20,27 @@
...
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
 import io.vertx.mqtt.messages.MqttPublishMessage;
 
 /**
  * Represents an MQTT PUBLISH message
  */
 public class MqttPublishMessageImpl implements MqttPublishMessage {
 
+  private static final Logger log = LoggerFactory.getLogger(MqttPublishMessageImpl.class);
```
[the remark this answers is not in the document]
Comment (dometec) [reply src/main/java/io/vertx/mqtt/messages/impl/MqttPublishMessageImpl.java]: right, I switched from log to throwing exception...

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by vietj** [comment]
thanks for the contribution @dometec

**Comment by dometec** [comment]
it was a pleasure ;)

**Pull request merged**